netket/logging/hdf5_log.py

```diff
@@ -0,0 +1,243 @@
+    keys = (obj.name,)
+    if isinstance(obj, h5py.Group):
+        for key, value in obj.items():
+            if isinstance(value, h5py.Group):
+                keys = keys + _allkeys(value)
+            else:
+                keys = keys + (value.name,)
+    return keys
+
+def _exists_hdf5(prefix):
+    return _path.exists(prefix + ".hdf5")
+
+_mode_shorthands = {
+    'write': 'w',
+    'append': 'a',
+    'fail': 'x'
+}
+
+class HDF5Log(RuntimeLog):
+    """
+    HDF5 Logger, that can be passed with keyword argument `logger` to Monte
+    Carlo drivers in order to serialize the output data of the simulation.
+
+    The logger has support for scalar numbers, NumPy/JAX arrays, and netket.stats.Stats objects.
+    These are stored as individual groups within a HDF5 file, under the main group `data/`:
+
+    - scalars are stored as a group with one dataset values of shape (n_steps,) containing the logged values,
+    - arrays are stored in the same way, but with values having shape (n_steps, *array_shape),
+    - netket.stats.Stats are stored as a group containing each field (Mean, Variance, etc...) as a separate dataset.
+
+    Importantly, each group has a dataset `iters`, which tracks the iteration number of the logged quantity.  
+
+    If the model state is serialized, then it is serialized as a dataset in the group `parameters/`.
+    The target of the serialization are the parameters of the variational state,
+    which are stored in a dataset `parameters/value` as a flattened array.
+
+    Data can be deserialized by calling :code:`f = h5py.File(filename, 'r')` and
+    inspecting the datasets as a dictionary, i.e. :code:`f['data/energy/Mean']`
+    """
+
+    def __init__(
+        self,
+        output_prefix: str,
+        mode: str = "write",
+        save_params_every: int = 50,
+        write_every: int = 50,
+        save_params: bool = True,
+        autoflush_cost: float = 0.005,
+    ):
+        """
+        Construct a HDF5 Logger.
+
+        Args:
+            output_prefix: the name of the output files before the extension
+            save_params_every: every how many iterations should machine parameters be
+                flushed to file
+            write_every: every how many iterations should data be flushed to file
+            mode: Specify the behaviour in case the file already exists at this
+                output_prefix. Options are
+                - `[w]rite`: (default) overwrites file if it already exists;
+                - `[x]` or `fail`: fails if file already exists;
+            save_params: bool flag indicating whether variables of the variational state
+                should be serialized at some interval. The output file is overwritten
+                every time variables are saved again.
+            autoflush_cost: Maximum fraction of runtime that can be dedicated to
+                serializing data. Defaults to 0.005 (0.5 per cent)
+        """
+        super().__init__()
```

Comment:
first thing, try to load h5py. If this fails it will print a nice error.
```suggestion
        import h5py

        super().__init__()
```